[SPARK-27001][SQL] Refactor "serializerFor" method between ScalaReflection and JavaTypeInference

HeartSaVioR opened the pull request:
## What changes were proposed in this pull request?

This patch proposes refactoring `serializerFor` method between `ScalaReflection` and `JavaTypeInference`, being consistent with what we refactored for `deserializerFor` in #23854.

This patch also extracts the logic on recording walk type path since the logic is duplicated across `serializerFor` and `deserializerFor` with `ScalaReflection` and `JavaTypeInference`.

## How was this patch tested?

Existing tests.

HeartSaVioR committed:
[SPARK-27001][SQL] Refactor "serializerFor" method between ScalaReflection and JavaTypeInference

HeartSaVioR committed:
Also extract the logic on recording walked type path

HeartSaVioR committed:
Address UDT as well

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #102825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102825/testReport)** for PR 23908 at commit [`9d62bc9`](https://github.com/apache/spark/commit/9d62bc92148d87427ef66e5eae463d72c6525a92).

SparkQA commented:
**[Test build #102825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102825/testReport)** for PR 23908 at commit [`9d62bc9`](https://github.com/apache/spark/commit/9d62bc92148d87427ef66e5eae463d72c6525a92).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102825/
Test FAILed.

HeartSaVioR commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePathRecorder.scala:
```diff
@@ -0,0 +1,49 @@
+
+package org.apache.spark.sql.catalyst
+
+object WalkedTypePathRecorder {
```
I was thinking about making this as class (with storing built path in each instance) but soon realized it requires touching other thing as well and feel a bit overkill. I'm still open to make this as individual class so please let me know if it sounds better.

cloud-fan replied:
making it a class looks better, as it needs to accumulate the walked type path.

HeartSaVioR replied:
Thanks for the support! Just addressed.

SparkQA commented:
**[Test build #102828 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102828/testReport)** for PR 23908 at commit [`3e17117`](https://github.com/apache/spark/commit/3e1711731f122444a0ff4999610c295f1364f414).

SparkQA commented:
**[Test build #102829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102829/testReport)** for PR 23908 at commit [`d7b8292`](https://github.com/apache/spark/commit/d7b82927aa5ead2569caa68b43ac2dfa7bdb633e).

fottey commented:
As an outside observer, would this refactoring allow the method `ScalaReflection.serializeFor` to handle arbitrary types that conform to the Java bean interface, and/or common Java specific types, such as `java.util.List`?

I recently discovered that because most of the common Scala implicit encoders reduce to `ExpressionEncoder`'s `apply` method, it's very difficult to work with arbitrary Java bean type's in the Dataset API.  

Specifically, given a java bean type, `MyBean`, and an implicit encoder of that bean type in scope, existing Spark 2.4.0 machinery in can't synthesize a valid encoder at runtime for hybrid Scala / Java types, like `Seq[MyBean]` or tuple types like `(Int, MyBean)` despite the fact that we have encoders for `Seq[_]`, `Tuple2[_, _]`, and `MyBean` available separately.

While it may be unreasonable to solve the problem generically across all potential classes, it would be really nice if `ExpressionEncoder`'s `apply` method could somehow detect and support at least java beans and java.util.Lists at runtime... 

See below code examples:

```scala
import com.example.MyBean
import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder

object Example {
    case class Test()

    def main(args: Array[String]): Unit = {
        val spark: SparkSession = ???

        import spark.implicits._

        // Works today after above implicit import
        val ds: Dataset[Seq[Test]] = Seq(Seq(Test()), Seq(Test()), ...).toDS

        // DOES NOT WORK
        // ExpressionEncoder's apply method cannot handle type MyBean!
        implicit def newMyBeanExpressionEncoder: Encoder[MyBean] = ExpressionEncoder()
        // 
        // Need to do the following:
        implicit def newMyBeanBeanEncoder: Encoder[MyBean] = Encoders.bean(classOf[MyBean])

        // But this only allows expressing things like this:
        val ds: Dataset[MyBean] = Seq(new MyBean(), new MyBean(), ...).toDS

        // Due to the above limitation we CANNOT do the following, EVEN AFTER
        // newMyBeanBeanEncoder is brought into scope!
        // DOES NOT WORK 
        val ds: Dataset[Seq[MyBean]] = Seq(Seq(new MyBean()), Seq(new MyBean()), ...).toDS

        // Finally, these do not work: 

        // DOES NOT WORK 
        val ds: Dataset[(Int, MyBean)] = Seq((0, new MyBean()),(0, new MyBean()), ...).toDS

        // DOES NOT WORK
        implicit def newMyBeanEncoder: Encoder[Seq[MyBean]] = ExpressionEncoder()
        
        // DOES NOT WORK
        implicit def newMyBeanEncoder: Encoder[java.util.List[MyBean]] = ExpressionEncoder()

        // The above samples all rely on ExpressionEncoder
        // being able to handle every type in the expression...
        // currently seems to work for:
        // - case classes
        // - tuples
        // - scala.Product
        // - scala "primitives"
        // other common types with encoders... BUT NOT java beans... :'(
    }
}
```

SparkQA commented:
**[Test build #102828 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102828/testReport)** for PR 23908 at commit [`3e17117`](https://github.com/apache/spark/commit/3e1711731f122444a0ff4999610c295f1364f414).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102828/
Test FAILed.

SparkQA commented:
**[Test build #102829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102829/testReport)** for PR 23908 at commit [`d7b8292`](https://github.com/apache/spark/commit/d7b82927aa5ead2569caa68b43ac2dfa7bdb633e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102829/
Test FAILed.

HeartSaVioR committed:
Fix a silly mistake

HeartSaVioR commented:
@fottey 
Could you file an issue for your case? Refactor is normally done without touching behavior, or at most making some minor changes, so I think it's out of scope for this PR.

HeartSaVioR committed:
Introduce WalkedTypePath class to replace recording walked path in `Seq[String]`

SparkQA commented:
**[Test build #102833 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102833/testReport)** for PR 23908 at commit [`01b7c41`](https://github.com/apache/spark/commit/01b7c4166aacab7a0485976b75fb8210993cdbd9).

SparkQA commented:
**[Test build #102833 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102833/testReport)** for PR 23908 at commit [`01b7c41`](https://github.com/apache/spark/commit/01b7c4166aacab7a0485976b75fb8210993cdbd9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102833/
Test FAILed.

HeartSaVioR committed:
Fixed string issue

HeartSaVioR commented:
Some test failures are occurred just because of one more new line. Fixed.
Some other test failures complain plan do not match: though string representation of two plans look same. Need to investigate.

SparkQA commented:
**[Test build #102841 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102841/testReport)** for PR 23908 at commit [`0d42fb4`](https://github.com/apache/spark/commit/0d42fb4244f7bcf0c517da4299ef882046a841a2).

HeartSaVioR committed:
Change WalkedTypePath to case class so that `equals` are properly implemented

HeartSaVioR commented:
The default implementation of `equals` in WalkedTypePath affects comparison. Fixed.

SparkQA commented:
**[Test build #102842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102842/testReport)** for PR 23908 at commit [`8e62f4c`](https://github.com/apache/spark/commit/8e62f4c2f1892c952fb7a97ebf7410df6231653f).

SparkQA commented:
**[Test build #102841 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102841/testReport)** for PR 23908 at commit [`0d42fb4`](https://github.com/apache/spark/commit/0d42fb4244f7bcf0c517da4299ef882046a841a2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102841/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:
```diff
@@ -48,8 +48,8 @@ object DeserializerBuildHelper {
-      funcForCreatingNewExpr: (Expression, Seq[String]) => Expression): Expression = {
+      walkedTypePath: WalkedTypePath,
+      funcForCreatingNewExpr: (Expression, WalkedTypePath) => Expression): Expression = {
     val newExpr = funcForCreatingNewExpr(expr, walkedTypePath)
```
this belongs to the previous PR: why not just let the caller side create the expression and pass to `deserializerForWithNullSafety`?

HeartSaVioR replied:
Yeah I might be considered too complicated. Not a big deal and looks simpler. Will address. Thanks!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -355,7 +352,7 @@ object JavaTypeInference {
   def serializerFor(beanClass: Class[_]): Expression = {
     val inputObject = BoundReference(0, ObjectType(beanClass), nullable = true)
-    val nullSafeInput = AssertNotNull(inputObject, Seq("top level input bean"))
+    val nullSafeInput = AssertNotNull(inputObject, new WalkedTypePath(Seq("top level input bean")))
```
why not use `recordRoot` here?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -364,65 +361,35 @@ object JavaTypeInference {
-          classOf[GenericArrayData],
-          input :: Nil,
-          dataType = ArrayType(dataType, nullable))
+        createSerializerForGenericArray(input, dataType, nullable = nullable)
```
`createSerializerForPrimitiveArray`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala:
```diff
@@ -0,0 +1,50 @@
+
+package org.apache.spark.sql.catalyst
+
+case class WalkedTypePath(walkedPaths: Seq[String] = Nil) extends Serializable {
```
shall we use a mutable list for better performance?

HeartSaVioR replied:
It was to address diverged paths for map key and value, but we can also copy instance via cloning internal list if necessary. Will address.

HeartSaVioR replied:
Addressed via 90df8a3 - I found it a bit complicated to maintain the list without polluting, so please take a look at the change and let me know if you would like to roll back to immutable one if performance gain doesn't seem to have more value than complexity.

SparkQA commented:
**[Test build #102842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102842/testReport)** for PR 23908 at commit [`8e62f4c`](https://github.com/apache/spark/commit/8e62f4c2f1892c952fb7a97ebf7410df6231653f).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class WalkedTypePath(walkedPaths: Seq[String] = Nil) extends Serializable `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102842/
Test FAILed.

HeartSaVioR committed:
address review comments from cloud-fan

- replace function parameter from function which creates instance to actual instance if it doesn't really necessary
- use recordRoot for initializing WalkedTypePath
- use mutable list for WalkedTypePath

SparkQA commented:
**[Test build #102850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102850/testReport)** for PR 23908 at commit [`90df8a3`](https://github.com/apache/spark/commit/90df8a3ebf3a228e4ff6d47bfd6f0ed98ad2b964).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:
```diff
@@ -39,40 +39,38 @@ object DeserializerBuildHelper {
     upCastToExpectedType(newPath, dataType, walkedTypePath)
   }
 
   def deserializerForWithNullSafety(
```
so this is same as `expressionWithNullSafety`?

HeartSaVioR replied:
Yeah now it's same. I've left comment on #23916. Btw I feel we might be better to focus review channel to one, either this PR or #23916 .

viirya replied:
I didn't notice you also make a lot change to `DeserializerBuildHelper` in this PR.  There might be conflicts if continuing #23916, I will close it.

HeartSaVioR replied:
I think `expressionWithNullSafety` is more general naming so might be preferred one, but `deserializerForWithNullSafety` is also a good name cause we have relevant method `deserializerForWithNullSafetyAndUpcast`.

So that's a matter of preference and either can be removed. Which method would we prefer to keep?

HeartSaVioR replied:
Thanks @viirya , please feel free to comment even it belongs to previous PR. Thanks again!

viirya replied:
Why you remove the `funcForCreatingNewExpr` from this and turn to pass in created expression (deserializer)?

HeartSaVioR replied:
I tend to follow the suggestion (https://github.com/apache/spark/pull/23908#discussion_r261019295) unless I have strong opinion, as I'm fairly new to contributing SQL area. For consistency I agree having func is better, but for simplicity we can remove it like applying inline. Either is reasonable.

cloud-fan replied:
I think we just need to keep `expressionWithNullSafety`. I don't see why we have to have 2 methods for `deserializeFor`. Leaving only a `deserializerForWithNullSafetyAndUpcast` is fine.

HeartSaVioR replied:
Left `deserializerForWithNullSafetyAndUpcast` and `expressionWithNullSafety` since both are used in multiple places. Please let me know if it doesn't work.

viirya commented:
So this is a purely refactoring PR and doesn't address bug, right?

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala:
```diff
@@ -0,0 +1,189 @@
+
+  def createSerializerForDouble(inputObject: Expression): Expression = {
+    Invoke(inputObject, "doubleValue", DoubleType)
+  }
```
Seems these helper methods don't reduce code and just add one more wrapper around calling `Invoke`. Are they needed?

HeartSaVioR replied:
This is not for reducing code. This is for consistency. These methods ensure we are consistently serialize / deserialize things between ScalaReflection and JavaTypeInference if the type is same.

HeartSaVioR commented:
> So this is a purely refactoring PR and doesn't address bug, right?

Yes, and make things consistent between `deserializerFor` and `serializerFor`.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -319,28 +323,26 @@ object ScalaReflection extends ScalaReflection {
-            path,
+            newPath,
             dataType,
             nullable = nullable,
```
Doesn't this revert what you did in previous PR?

HeartSaVioR replied:
I got review comment for previous PR as well - this is the reflection of replacing function to applied expression which ends up making `deserializerForWithNullSafety` and `expressionWithNullSafety` being same.

SparkQA commented:
**[Test build #102850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102850/testReport)** for PR 23908 at commit [`90df8a3`](https://github.com/apache/spark/commit/90df8a3ebf3a228e4ff6d47bfd6f0ed98ad2b964).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class WalkedTypePath() extends Serializable `
  * `case class AssertNotNull(child: Expression, walkedTypePath: WalkedTypePath = WalkedTypePath())`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102850/
Test FAILed.

HeartSaVioR committed:
Remove unnecessary method

SparkQA commented:
**[Test build #102882 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102882/testReport)** for PR 23908 at commit [`ff7512b`](https://github.com/apache/spark/commit/ff7512bc3aead2458b1fdaf3ea1609373ade9da7).

SparkQA commented:
**[Test build #102882 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102882/testReport)** for PR 23908 at commit [`ff7512b`](https://github.com/apache/spark/commit/ff7512bc3aead2458b1fdaf3ea1609373ade9da7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102882/
Test PASSed.

HeartSaVioR commented:
Rebased against latest master to incorporate #23913.

Btw, I've realized the boundary of refactor can be expanded to RowEncoder, for both `serializerFor` and `deserializerFor`. Would it be better to address it with new PR, or is it OK to address here as well?

SparkQA commented:
**[Test build #102899 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102899/testReport)** for PR 23908 at commit [`852debd`](https://github.com/apache/spark/commit/852debda5c634f12ddacda42bab4d0f69596b753).

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala:
```diff
@@ -0,0 +1,63 @@
+import scala.collection.mutable
+
+case class WalkedTypePath() extends Serializable {
+  val walkedPaths: mutable.ArrayBuffer[String] = new mutable.ArrayBuffer[String]()
```
private?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala:
```diff
@@ -0,0 +1,63 @@
+
+import scala.collection.mutable
+
+case class WalkedTypePath() extends Serializable {
```
We need to make this serializable? It seems this class is used only in a driver side?

HeartSaVioR replied:
It was non-serializable and encountered error on existing tests so added it. Unfortunately I don't remember which one was.

maropu replied:
To check which test fails, I opened a pr as followup: #23973 (I'm a bit interested in this...).

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala:
```diff
@@ -0,0 +1,63 @@
+package org.apache.spark.sql.catalyst
+
+import scala.collection.mutable
+
```
Probably, we would be better to describe what this class is used for.

maropu commented:
Nice work, @HeartSaVioR! btw, this pr consists of the two parts you described in the PR description. If so, how about splitting this into the two prs for easy reviews? Refactoring the code for the consistency between ScalaReflection and JavaTypeInference, and adding WalkedTypePath then?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:
```diff
@@ -39,40 +39,30 @@ object DeserializerBuildHelper {
       expr
     } else {
-      AssertNotNull(expr, walkedTypePath)
+      AssertNotNull(expr, walkedTypePath.copy())
```
We can let `AssertNotNull` take a `Seq[String]`, to force us to copy the `WalkedTypePath` when creating `AssertNotNull`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:
```diff
@@ -167,10 +157,10 @@ object DeserializerBuildHelper {
     case _: ArrayType => expr
     case _: MapType => expr
-    case _ => UpCast(expr, expected, walkedTypePath)
+    case _ => UpCast(expr, expected, walkedTypePath.copy())
```
ditto

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -274,38 +274,40 @@ object JavaTypeInference {
+          valueType.getType.getTypeName)
+
+        val newTypePathForKey = walkedTypePath.copy()
+        val newTypePathForValue = walkedTypePath.copy()
```
Sorry for the back and forth. But seems it's better to make `WalkedTypePath` immutable as there are branches. It's hard to maintain and we can easily mess it up if we forget the call `copy` somewhere.

HeartSaVioR replied:
Yeah, same understanding. I'll revert back to let WalkedTypePath be immutable one.

HeartSaVioR commented:
@maropu I like the idea of splitting PR, but since @cloud-fan already provides feedbacks on WalkedTypePath, might be better to hear opinion and decide. Let me first address his feedback on WalkedTypePath - even we decide to break down it would be needed work.

HeartSaVioR committed:
Revert "address review comments from cloud-fan"

This reverts commit c67826afb738bfe5b74447ce91e3e6ade9f9a484.

NOTE: there's conflict which makes revert commit not clearly reverting as before, but WalkedTypePath is clearly reverted

HeartSaVioR committed:
Address review comment from cloud-fan (partially reapplying 90df8a3ebf3a228e4ff6d47bfd6f0ed98ad2b964)

HeartSaVioR committed:
Address review comments from maropu

cloud-fan commented:
I'm OK both ways. Since the PR already contains the `WalkedTypePath` refactor, I think it's fine to include it as it's pretty intuitive.

HeartSaVioR commented:
Thanks! I'll keep it as it is. How about applying `DeserializerBuildHelper`/`SerializerBuildHelper` to RowEncoder? Better to have it as separate PR?

SparkQA commented:
**[Test build #102902 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102902/testReport)** for PR 23908 at commit [`578d8fe`](https://github.com/apache/spark/commit/578d8feff26565f7ee6a0be5ff85a244ff94d873).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -387,38 +383,28 @@ object ScalaReflection extends ScalaReflection {
-              "fromPrimitiveArray",
-              input :: Nil,
-              returnNullable = false)
+            createSerializerForPrimitiveArray(input, dt)
```
seems this branch is missing in the java side. We can address it in the followup.

HeartSaVioR replied:
Raised PR: https://github.com/apache/spark/pull/24015

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:
```diff
@@ -1378,7 +1378,10 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
+case class UpCast(
+    child: Expression,
+    dataType: DataType,
+    walkedTypePath: WalkedTypePath = new WalkedTypePath())
```
can we keep it `Seq[String]`? When we reach here, the `walkedTypePath` is only needed for logging/error message, and we don't need the `WalkedTypePath` class to help accumulate the paths.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:
```diff
@@ -1627,7 +1627,7 @@ case class InitializeJavaBean(beanInstance: Expression, setters: Map[String, Exp
  * non-null `s`, `s.i` can't be null.
  */
-case class AssertNotNull(child: Expression, walkedTypePath: Seq[String] = Nil)
+case class AssertNotNull(child: Expression, walkedTypePath: WalkedTypePath = new WalkedTypePath())
```
ditto

SparkQA commented:
**[Test build #102899 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102899/testReport)** for PR 23908 at commit [`852debd`](https://github.com/apache/spark/commit/852debda5c634f12ddacda42bab4d0f69596b753).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102899/
Test PASSed.

HeartSaVioR committed:
Address review comments

SparkQA commented:
**[Test build #102906 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102906/testReport)** for PR 23908 at commit [`20e8d5a`](https://github.com/apache/spark/commit/20e8d5a0b4608d3b0ccd4fe00553065d20b2b6fe).

SparkQA commented:
**[Test build #102902 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102902/testReport)** for PR 23908 at commit [`578d8fe`](https://github.com/apache/spark/commit/578d8feff26565f7ee6a0be5ff85a244ff94d873).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102902/
Test PASSed.

SparkQA commented:
**[Test build #102906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102906/testReport)** for PR 23908 at commit [`20e8d5a`](https://github.com/apache/spark/commit/20e8d5a0b4608d3b0ccd4fe00553065d20b2b6fe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class AssertNotNull(child: Expression, walkedTypePath: Seq[String] = Nil)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102906/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -358,7 +353,8 @@ object JavaTypeInference {
     val inputObject = BoundReference(0, ObjectType(beanClass), nullable = true)
-    val nullSafeInput = AssertNotNull(inputObject, Seq("top level input bean"))
+    val nullSafeInput = AssertNotNull(inputObject,
+      WalkedTypePath().recordRoot("top level input bean").getPaths)
```
nit: we can keep it unchanged.

HeartSaVioR replied:
Ah yes that's not even same. Will revert.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:
```diff
@@ -195,7 +195,7 @@ case class ExpressionEncoder[T](
           // For input object of Product type, we can't encode it to row if it's null, as Spark SQL
           // doesn't allow top-level row to be null, only its columns can be null.
-          AssertNotNull(r, Seq("top level Product or row object"))
+          AssertNotNull(r, WalkedTypePath().recordRoot("top level Product or row object").getPaths)
```
nit: we can keep it unchanged.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:
```diff
@@ -1378,7 +1378,10 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
+case class UpCast(
+    child: Expression,
+    dataType: DataType,
+    walkedTypePath: Seq[String] = Nil)
```
can we revert the code style change?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:
```diff
@@ -28,7 +28,7 @@ import scala.util.Try
 import org.apache.spark.serializer._
 import org.apache.spark.sql.Row
-import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow, ScalaReflection}
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow, ScalaReflection, WalkedTypePath}
```
unnecessary change

HeartSaVioR committed:
Address review comments

SparkQA commented:
**[Test build #102936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102936/testReport)** for PR 23908 at commit [`50c2ddc`](https://github.com/apache/spark/commit/50c2ddce1e527b0d46d5ac1fbc2de28575d4059e).

SparkQA commented:
**[Test build #102936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102936/testReport)** for PR 23908 at commit [`50c2ddc`](https://github.com/apache/spark/commit/50c2ddce1e527b0d46d5ac1fbc2de28575d4059e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UpCast(child: Expression, dataType: DataType, walkedTypePath: Seq[String] = Nil)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102936/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

HeartSaVioR commented:
Thanks all for reviewing and merging!
